File: aic_data_collection/test/test_completion_monitor.py
import pytest
import math

from aic_data_collection.completion_monitor import CompletionResult, check_tf_completion


class TestCheckTfCompletion:
    def test_success_within_threshold(self):
        plug_pos = (0.1, 0.2, 0.3)
        port_pos = (0.1, 0.2, 0.303)  # 3mm away
        plug_quat = (0.0, 0.0, 0.0, 1.0)
        port_quat = (0.0, 0.0, 0.0, 1.0)
        result = check_tf_completion(plug_pos, port_pos, plug_quat, port_quat,
                                     distance_threshold=0.005, orientation_threshold=0.1)
        assert result is True

    def test_fail_distance_too_large(self):
        plug_pos = (0.1, 0.2, 0.3)
        port_pos = (0.1, 0.2, 0.31)  # 10mm away
        plug_quat = (0.0, 0.0, 0.0, 1.0)
        port_quat = (0.0, 0.0, 0.0, 1.0)
        result = check_tf_completion(plug_pos, port_pos, plug_quat, port_quat,
                                     distance_threshold=0.005, orientation_threshold=0.1)
        assert result is False

    def test_fail_orientation_too_large(self):
        plug_pos = (0.1, 0.2, 0.3)
        port_pos = (0.1, 0.2, 0.301)
        plug_quat = (0.0, 0.0, 0.0, 1.0)
        # ~0.2 rad rotation around z
        port_quat = (0.0, 0.0, math.sin(0.1), math.cos(0.1))
        result = check_tf_completion(plug_pos, port_pos, plug_quat, port_quat,
                                     distance_threshold=0.005, orientation_threshold=0.1)
        assert result is False


class TestCompletionResult:
    def test_success(self):
        r = CompletionResult.SUCCESS
        assert r.is_success()

    def test_partial_is_success(self):
        r = CompletionResult.PARTIAL
        assert r.is_success()

    def test_timeout_is_not_success(self):
        r = CompletionResult.TIMEOUT
        assert not r.is_success()

    def test_action_failed_is_not_success(self):
        r = CompletionResult.ACTION_FAILED
        assert not r.is_success()

File: aic_engine/test/test_completion_monitor.cpp
#include <gtest/gtest.h>
#include <cmath>
#include "completion_monitor.hpp"

using aic::CompletionResult;
using aic::check_tf_completion;

TEST(CheckTfCompletion, SuccessWithinThreshold) {
  Eigen::Vector3d plug_pos(0.1, 0.2, 0.3);
  Eigen::Vector3d port_pos(0.1, 0.2, 0.303);
  Eigen::Quaterniond plug_quat(1.0, 0.0, 0.0, 0.0);
  Eigen::Quaterniond port_quat(1.0, 0.0, 0.0, 0.0);
  EXPECT_TRUE(check_tf_completion(plug_pos, port_pos, plug_quat, port_quat, 0.005, 0.1));
}

TEST(CheckTfCompletion, FailDistanceTooLarge) {
  Eigen::Vector3d plug_pos(0.1, 0.2, 0.3);
  Eigen::Vector3d port_pos(0.1, 0.2, 0.31);
  Eigen::Quaterniond plug_quat(1.0, 0.0, 0.0, 0.0);
  Eigen::Quaterniond port_quat(1.0, 0.0, 0.0, 0.0);
  EXPECT_FALSE(check_tf_completion(plug_pos, port_pos, plug_quat, port_quat, 0.005, 0.1));
}

TEST(CheckTfCompletion, FailOrientationTooLarge) {
  Eigen::Vector3d plug_pos(0.1, 0.2, 0.3);
  Eigen::Vector3d port_pos(0.1, 0.2, 0.301);
  Eigen::Quaterniond plug_quat(1.0, 0.0, 0.0, 0.0);
  Eigen::Quaterniond port_quat(std::cos(0.1), 0.0, 0.0, std::sin(0.1));
  EXPECT_FALSE(check_tf_completion(plug_pos, port_pos, plug_quat, port_quat, 0.005, 0.1));
}

TEST(CompletionResult, SuccessIsSuccess) {
  EXPECT_TRUE(aic::is_success(CompletionResult::SUCCESS));
}

TEST(CompletionResult, PartialIsSuccess) {
  EXPECT_TRUE(aic::is_success(CompletionResult::PARTIAL));
}

TEST(CompletionResult, TimeoutIsNotSuccess) {
  EXPECT_FALSE(aic::is_success(CompletionResult::TIMEOUT));
}

TEST(CompletionResult, ActionFailedIsNotSuccess) {
  EXPECT_FALSE(aic::is_success(CompletionResult::ACTION_FAILED));
}
